hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -139,8 +139,10 @@ object InferPType {
     case x: StreamScan => x.accPType
     case StreamFold2(a, _, `name`, _, _) => coerce[PStream](a.pType).elementType
     case x: StreamFold2 => x.accPTypes(x.nameIdx(name))
-    case StreamLeftJoinDistinct(left, _, `name`, _, _, _) => coerce[PStream](left.pType).elementType
-    case StreamLeftJoinDistinct(_, right, _, `name`, _, _) => coerce[PStream](right.pType).elementType.setRequired(false)
+    case StreamJoinRightDistinct(left, _, _, _, `name`, _, _, joinType) =>
+      coerce[PStream](left.pType).elementType.setRequired(joinType == "left")
```

Comment:
if the left element is optional, this can still return required = true. Should probably use `setMissing(joinType != "left")`

Reply:
Oops, good catch. That's what I had before the rebase moved requiredness logic around.